src/main/java/org/jitsi/videobridge/octo/OctoTransportManager.java

```diff
@@ -0,0 +1,459 @@
+
+    /**
+     * Sets the list of IDs of remote relays, i.e. the targets that this
+     * {@link TransportManager} will send packets to.
+     */
+    public void setRelayIds(List<String> relayIds)
+    {
+        List<SocketAddress> remoteRelays = new ArrayList<>(relayIds.size());
+
+        for (String relayId : relayIds)
+        {
+            SocketAddress socketAddress = relayIdToSocketAddress(relayId);
+            if (socketAddress == null)
+            {
+                logger.error(
+                    "Could not convert a relay ID to a socket address: "
+                        + relayId);
+                continue;
+            }
+
+            if (remoteRelays.contains(socketAddress))
+            {
+                logger.info("Relay ID duplicate: " + relayId);
+                continue;
+            }
+
+            remoteRelays.add(socketAddress);
+        }
+
+        this.remoteRelays = remoteRelays;
+    }
+
+    /**
+     * Sends a specific {@link DatagramPacket} to all targets of this
+     * {@link TransportManager} (i.e. to all {@link #remoteRelays}).
+     * The payload of the packet is first encapsulated in Octo (i.e. the fixed
+     * size Octo header is prepended).
+     *
+     * @param p the packet to send.
+     * @throws IOException if calling {@code send()} on any of the target
+     * sockets results in an {@link IOException}.
+     */
+    private void doSend(DatagramPacket p)
+        throws IOException
+    {
+        SocketAddress originalAddress = p.getSocketAddress();
+        p = addOctoHeaders(p);
+        DatagramSocket relaySocket = octoRelay.getSocket();
+        IOException exception = null;
+        int exceptions = 0;
+
+        for (SocketAddress remoteAddress : remoteRelays)
+        {
+            try
+            {
+                p.setSocketAddress(remoteAddress);
+
+                relaySocket.send(p);
+            }
+            catch (IOException ioe)
+            {
+                exceptions++;
+                exception = ioe;
+            }
+        }
+
+        // Restore the original address. Note that we don't currently restore
+        // the payload (i.e. we haven't removed the headers we added). Should
+        // we do that?
+        p.setSocketAddress(originalAddress);
+
+        if (exception != null)
+        {
+            logger.warn(
+                "Caught " + exceptions + " while trying to send a packet.");
+            throw exception;
+        }
+    }
+
+    /**
+     * Prepends the fixed-length Octo header to a packet.
+     * @param p the packet
+     * @return the packet to which an Octo header was added.
+     */
+    private DatagramPacket addOctoHeaders(DatagramPacket p)
+    {
+        byte[] buf = p.getData();
+        int off = p.getOffset();
+        int len = p.getLength();
+
+
+        if (off >= OctoPacket.OCTO_HEADER_LENGTH)
+        {
+            // If there is space before the offset, just use it.
+            off -= OctoPacket.OCTO_HEADER_LENGTH;
```

Comment:
this seems a bit weird to me: is the offset not basically communicating "as far as you're concerned, the buffer starts here"?  should we really be writing over data before the offset?  i think this should trust/uphold the contract of the buf/len/offset fields.

Reply:
Hmm. This is interesting. It would be unfortunate if can't use this, as it might affect performance. I'm pretty sure we already do something like this in other places, so I think in our use-case it's safe. I'll do some digging and perhaps we can talk about this offline too

Reply:
It looks like DatagramSocket#receive will respect the offset and not change it or override the data at smaller indexes. However in our code, we generally use a given byte[] only for one packet at a time. That is, there are never two packets sharing a byte[], but a byte[] can be recycled and reused for another packet. I think in this case it's a reasonable assumption that the whole byte[] is available to the packet (it just happens to be that there are some unused bytes if offset>0), and therefore the offset can be reset. The code above works with DatagramPackets which come from RTPConnectorOutputStream, and the assumption holds.

Reply:
to me this seems like a "gotcha" waiting to happen...we pass buf, off, len in lots of places under the premise that they will be used as documented (what would the purpose of passing `off` be otherwise) but now we have a place that ignores the `off` field

Reply:
`off` isn't ignored, it is updated. In this scheme, its purpose is to indicate where the packet's data starts (and not to protect the memory before `off`).

Reply:
ok, i think after re-reading this some more (and also noticing that it's actually a `DatagramPacket` that's passed and not individual buf/off/len fields) i think i agree that it feels ok.